[Automation] java run fails for org.jline:jline-terminal-jni:3.30.7

## java run on ubuntu-latest / JDK 25 / mode current-defaults

Failure kind: java run
OS: ubuntu-latest
JDK: 25
Native-image mode: current-defaults
Reproducer: `GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="3.30.7" ./gradlew clean javaTest -Pcoordinates="org.jline:jline-terminal-jni:3.27.1"`
Runner log: https://github.com/oracle/graalvm-reachability-metadata/actions/runs/25296069081/job/74155118880
Last 300 lines of the log:
```console
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE

> Configure project :
GraalVM Reachability Metadata TCK
---------------------------------

> Task :tckClean
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :clean

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.jline/jline-terminal-jni/3.27.1/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 756ms
9 actionable tasks: 1 executed, 8 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

> Task :clean

> Task :compileTestJava
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :compileJava NO-SOURCE
> Task :processResources NO-SOURCE
> Task :classes UP-TO-DATE
> Task :tck-test-support:compileJava UP-TO-DATE
> Task :compileTestJava

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.jline/jline-terminal-jni/3.27.1/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 934ms
10 actionable tasks: 1 executed, 9 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 2s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'javac compile' for 3.30.7 at 2026-05-04T01:19:53Z with exit code 0 after 3s
========================================================================================
 org.jline:jline-terminal-jni:3.27.1:3.30.7 stage java run
========================================================================================
Starting stage 'java run' for 3.30.7 at 2026-05-04T01:19:53Z
Command: GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="3.30.7" ./gradlew clean javaTest -Pcoordinates="org.jline:jline-terminal-jni:3.27.1"
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE

> Configure project :
GraalVM Reachability Metadata TCK
---------------------------------

> Task :tckClean
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :clean

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.jline/jline-terminal-jni/3.27.1/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 743ms
9 actionable tasks: 1 executed, 8 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

> Task :clean

> Task :javaTest
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :checkTestTimeoutAnnotations
> Task :compileJava NO-SOURCE
> Task :tck-test-support:compileJava UP-TO-DATE
> Task :tck-test-support:processResources NO-SOURCE
> Task :processResources NO-SOURCE
> Task :classes UP-TO-DATE
> Task :tck-test-support:classes UP-TO-DATE
> Task :tck-test-support:jar UP-TO-DATE
> Task :compileTestJava
> Task :processTestResources NO-SOURCE
> Task :testClasses

> Task :test FAILED

Jline_terminal_jniTest > loadsJniTerminalProviderThroughPublicServiceApi() FAILED
    java.io.IOException: Error reading provider resource file: Unable to load terminal provider jni: null
        at org.jline.terminal.spi.TerminalProvider.load(TerminalProvider.java:498)
        at org_jline.jline_terminal_jni.Jline_terminal_jniTest.loadsJniTerminalProviderThroughPublicServiceApi(Jline_terminal_jniTest.java:42)

        Caused by:
        java.io.IOException: Unable to load terminal provider jni: null
            at org.jline.terminal.spi.TerminalProvider.load(TerminalProvider.java:493)
            ... 1 more

            Caused by:
            java.lang.reflect.InvocationTargetException
                at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:74)
                at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:502)
                at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:486)
                at org.jline.terminal.spi.TerminalProvider.load(TerminalProvider.java:491)
                ... 1 more

                Caused by:
                java.lang.UnsupportedOperationException: Native access is not enabled for the current module: unnamed module @1cb3ec38
                    at org.jline.terminal.impl.jni.JniTerminalProvider.checkNativeAccess(JniTerminalProvider.java:87)
                    at org.jline.terminal.impl.jni.JniTerminalProvider.<init>(JniTerminalProvider.java:67)
                    at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:62)
                    ... 4 more

Jline_terminal_jniTest > terminalBuilderCreatesJniBackedTerminalForExplicitStreams() FAILED
    java.lang.IllegalStateException: Unable to create a terminal
        at org.jline.terminal.TerminalBuilder.doBuild(TerminalBuilder.java:748)
        at org.jline.terminal.TerminalBuilder.build(TerminalBuilder.java:717)
        at org_jline.jline_terminal_jni.Jline_terminal_jniTest.buildTerminal(Jline_terminal_jniTest.java:221)
        at org_jline.jline_terminal_jni.Jline_terminal_jniTest.terminalBuilderCreatesJniBackedTerminalForExplicitStreams(Jline_terminal_jniTest.java:158)

Jline_terminal_jniTest > nativePtyAppliesAttributeChangesAfterOpen() FAILED
    java.io.IOException: Error reading provider resource file: Unable to load terminal provider jni: null
        at org.jline.terminal.spi.TerminalProvider.load(TerminalProvider.java:498)
        at org_jline.jline_terminal_jni.Jline_terminal_jniTest.nativePtyAppliesAttributeChangesAfterOpen(Jline_terminal_jniTest.java:126)

        Caused by:
        java.io.IOException: Unable to load terminal provider jni: null
            at org.jline.terminal.spi.TerminalProvider.load(TerminalProvider.java:493)
            ... 1 more

            Caused by:
            java.lang.reflect.InvocationTargetException
                at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:74)
                at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:502)
                at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:486)
                at org.jline.terminal.spi.TerminalProvider.load(TerminalProvider.java:491)
                ... 1 more

                Caused by:
                java.lang.UnsupportedOperationException: Native access is not enabled for the current module: unnamed module @1cb3ec38
                    at org.jline.terminal.impl.jni.JniTerminalProvider.checkNativeAccess(JniTerminalProvider.java:87)
                    at org.jline.terminal.impl.jni.JniTerminalProvider.<init>(JniTerminalProvider.java:67)
                    at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:62)
                    ... 4 more

Jline_terminal_jniTest > openCreatesConfiguredNativePtyOnSupportedPosixPlatforms() FAILED
    java.io.IOException: Error reading provider resource file: Unable to load terminal provider jni: null
        at org.jline.terminal.spi.TerminalProvider.load(TerminalProvider.java:498)
        at org_jline.jline_terminal_jni.Jline_terminal_jniTest.openCreatesConfiguredNativePtyOnSupportedPosixPlatforms(Jline_terminal_jniTest.java:56)

        Caused by:
        java.io.IOException: Unable to load terminal provider jni: null
            at org.jline.terminal.spi.TerminalProvider.load(TerminalProvider.java:493)
            ... 1 more

            Caused by:
            java.lang.reflect.InvocationTargetException
                at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:74)
                at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:502)
                at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:486)
                at org.jline.terminal.spi.TerminalProvider.load(TerminalProvider.java:491)
                ... 1 more

                Caused by:
                java.lang.UnsupportedOperationException: Native access is not enabled for the current module: unnamed module @1cb3ec38
                    at org.jline.terminal.impl.jni.JniTerminalProvider.checkNativeAccess(JniTerminalProvider.java:87)
                    at org.jline.terminal.impl.jni.JniTerminalProvider.<init>(JniTerminalProvider.java:67)
                    at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:62)
                    ... 4 more

Jline_terminal_jniTest > nativePtyExposesConnectedMasterAndSlaveStreams() FAILED
    java.io.IOException: Error reading provider resource file: Unable to load terminal provider jni: null
        at org.jline.terminal.spi.TerminalProvider.load(TerminalProvider.java:498)
        at org_jline.jline_terminal_jni.Jline_terminal_jniTest.nativePtyExposesConnectedMasterAndSlaveStreams(Jline_terminal_jniTest.java:95)

        Caused by:
        java.io.IOException: Unable to load terminal provider jni: null
            at org.jline.terminal.spi.TerminalProvider.load(TerminalProvider.java:493)
            ... 1 more

            Caused by:
            java.lang.reflect.InvocationTargetException
                at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:74)
                at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:502)
                at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:486)
                at org.jline.terminal.spi.TerminalProvider.load(TerminalProvider.java:491)
                ... 1 more

                Caused by:
                java.lang.UnsupportedOperationException: Native access is not enabled for the current module: unnamed module @1cb3ec38
                    at org.jline.terminal.impl.jni.JniTerminalProvider.checkNativeAccess(JniTerminalProvider.java:87)
                    at org.jline.terminal.impl.jni.JniTerminalProvider.<init>(JniTerminalProvider.java:67)
                    at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:62)
                    ... 4 more


5 tests completed, 5 failed

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':test'.
> There were failing tests. See the report at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.jline/jline-terminal-jni/3.27.1/build/reports/tests/test/index.html

* Try:
> Run with --scan to generate a Build Scan (Powered by Develocity).

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.jline/jline-terminal-jni/3.27.1/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.
13 actionable tasks: 3 executed, 10 up-to-date
BUILD FAILED in 2s

> Task :javaTest FAILED

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':javaTest'.
> Java tests failed

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to generate a Build Scan (Powered by Develocity).
> Get more help at https://help.gradle.org.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 3s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'java run' for 3.30.7 at 2026-05-04T01:19:57Z with exit code 1 after 4s
❌ ERROR: 'java run' for 3.30.7 failed with exit code 1.
FAILED[java run][3.30.7][GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="3.30.7" ./gradlew clean javaTest -Pcoordinates="org.jline:jline-terminal-jni:3.27.1"]
```

## java run on ubuntu-latest / JDK latest-ea / mode current-defaults

Failure kind: java run
OS: ubuntu-latest
JDK: latest-ea
Native-image mode: current-defaults
Reproducer: `GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="3.30.7" ./gradlew clean javaTest -Pcoordinates="org.jline:jline-terminal-jni:3.27.1"`
Runner log: https://github.com/oracle/graalvm-reachability-metadata/actions/runs/25296069081/job/74155118920
Last 300 lines of the log:
```console
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE

> Configure project :
GraalVM Reachability Metadata TCK
---------------------------------

> Task :tckClean
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :clean

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.jline/jline-terminal-jni/3.27.1/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 588ms
9 actionable tasks: 1 executed, 8 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

> Task :clean

> Task :compileTestJava
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :compileJava NO-SOURCE
> Task :tck-test-support:compileJava UP-TO-DATE
> Task :processResources NO-SOURCE
> Task :classes UP-TO-DATE
> Task :compileTestJava

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.jline/jline-terminal-jni/3.27.1/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 726ms
10 actionable tasks: 1 executed, 9 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 2s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'javac compile' for 3.30.7 at 2026-05-04T01:17:05Z with exit code 0 after 2s
========================================================================================
 org.jline:jline-terminal-jni:3.27.1:3.30.7 stage java run
========================================================================================
Starting stage 'java run' for 3.30.7 at 2026-05-04T01:17:05Z
Command: GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="3.30.7" ./gradlew clean javaTest -Pcoordinates="org.jline:jline-terminal-jni:3.27.1"
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE

> Configure project :
GraalVM Reachability Metadata TCK
---------------------------------

> Task :tckClean
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :clean

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.jline/jline-terminal-jni/3.27.1/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 594ms
9 actionable tasks: 1 executed, 8 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

> Task :clean

> Task :javaTest FAILED
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :checkTestTimeoutAnnotations
> Task :compileJava NO-SOURCE
> Task :tck-test-support:compileJava UP-TO-DATE
> Task :processResources NO-SOURCE
> Task :classes UP-TO-DATE
> Task :tck-test-support:processResources NO-SOURCE
> Task :tck-test-support:classes UP-TO-DATE
> Task :tck-test-support:jar UP-TO-DATE
> Task :compileTestJava
> Task :processTestResources NO-SOURCE
> Task :testClasses

> Task :test FAILED

Jline_terminal_jniTest > loadsJniTerminalProviderThroughPublicServiceApi() FAILED
    java.io.IOException: Error reading provider resource file: Unable to load terminal provider jni: null
        at org.jline.terminal.spi.TerminalProvider.load(TerminalProvider.java:498)
        at org_jline.jline_terminal_jni.Jline_terminal_jniTest.loadsJniTerminalProviderThroughPublicServiceApi(Jline_terminal_jniTest.java:42)

        Caused by:
        java.io.IOException: Unable to load terminal provider jni: null
            at org.jline.terminal.spi.TerminalProvider.load(TerminalProvider.java:493)
            ... 1 more

            Caused by:
            java.lang.reflect.InvocationTargetException
                at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:74)
                at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:502)
                at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:486)
                at org.jline.terminal.spi.TerminalProvider.load(TerminalProvider.java:491)
                ... 1 more

                Caused by:
                java.lang.UnsupportedOperationException: Native access is not enabled for the current module: unnamed module @1cb3ec38
                    at org.jline.terminal.impl.jni.JniTerminalProvider.checkNativeAccess(JniTerminalProvider.java:87)
                    at org.jline.terminal.impl.jni.JniTerminalProvider.<init>(JniTerminalProvider.java:67)
                    at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:62)
                    ... 4 more

Jline_terminal_jniTest > terminalBuilderCreatesJniBackedTerminalForExplicitStreams() FAILED
    java.lang.IllegalStateException: Unable to create a terminal
        at org.jline.terminal.TerminalBuilder.doBuild(TerminalBuilder.java:748)
        at org.jline.terminal.TerminalBuilder.build(TerminalBuilder.java:717)
        at org_jline.jline_terminal_jni.Jline_terminal_jniTest.buildTerminal(Jline_terminal_jniTest.java:221)
        at org_jline.jline_terminal_jni.Jline_terminal_jniTest.terminalBuilderCreatesJniBackedTerminalForExplicitStreams(Jline_terminal_jniTest.java:158)

Jline_terminal_jniTest > nativePtyAppliesAttributeChangesAfterOpen() FAILED
    java.io.IOException: Error reading provider resource file: Unable to load terminal provider jni: null
        at org.jline.terminal.spi.TerminalProvider.load(TerminalProvider.java:498)
        at org_jline.jline_terminal_jni.Jline_terminal_jniTest.nativePtyAppliesAttributeChangesAfterOpen(Jline_terminal_jniTest.java:126)

        Caused by:
        java.io.IOException: Unable to load terminal provider jni: null
            at org.jline.terminal.spi.TerminalProvider.load(TerminalProvider.java:493)
            ... 1 more

            Caused by:
            java.lang.reflect.InvocationTargetException
                at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:74)
                at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:502)
                at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:486)
                at org.jline.terminal.spi.TerminalProvider.load(TerminalProvider.java:491)
                ... 1 more

                Caused by:
                java.lang.UnsupportedOperationException: Native access is not enabled for the current module: unnamed module @1cb3ec38
                    at org.jline.terminal.impl.jni.JniTerminalProvider.checkNativeAccess(JniTerminalProvider.java:87)
                    at org.jline.terminal.impl.jni.JniTerminalProvider.<init>(JniTerminalProvider.java:67)
                    at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:62)
                    ... 4 more

Jline_terminal_jniTest > openCreatesConfiguredNativePtyOnSupportedPosixPlatforms() FAILED
    java.io.IOException: Error reading provider resource file: Unable to load terminal provider jni: null
        at org.jline.terminal.spi.TerminalProvider.load(TerminalProvider.java:498)
        at org_jline.jline_terminal_jni.Jline_terminal_jniTest.openCreatesConfiguredNativePtyOnSupportedPosixPlatforms(Jline_terminal_jniTest.java:56)

        Caused by:
        java.io.IOException: Unable to load terminal provider jni: null
            at org.jline.terminal.spi.TerminalProvider.load(TerminalProvider.java:493)
            ... 1 more

            Caused by:
            java.lang.reflect.InvocationTargetException
                at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:74)
                at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:502)
                at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:486)
                at org.jline.terminal.spi.TerminalProvider.load(TerminalProvider.java:491)
                ... 1 more

                Caused by:
                java.lang.UnsupportedOperationException: Native access is not enabled for the current module: unnamed module @1cb3ec38
                    at org.jline.terminal.impl.jni.JniTerminalProvider.checkNativeAccess(JniTerminalProvider.java:87)
                    at org.jline.terminal.impl.jni.JniTerminalProvider.<init>(JniTerminalProvider.java:67)
                    at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:62)
                    ... 4 more

Jline_terminal_jniTest > nativePtyExposesConnectedMasterAndSlaveStreams() FAILED
    java.io.IOException: Error reading provider resource file: Unable to load terminal provider jni: null
        at org.jline.terminal.spi.TerminalProvider.load(TerminalProvider.java:498)
        at org_jline.jline_terminal_jni.Jline_terminal_jniTest.nativePtyExposesConnectedMasterAndSlaveStreams(Jline_terminal_jniTest.java:95)

        Caused by:
        java.io.IOException: Unable to load terminal provider jni: null
            at org.jline.terminal.spi.TerminalProvider.load(TerminalProvider.java:493)
            ... 1 more

            Caused by:
            java.lang.reflect.InvocationTargetException
                at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:74)
                at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:502)
                at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:486)
                at org.jline.terminal.spi.TerminalProvider.load(TerminalProvider.java:491)
                ... 1 more

                Caused by:
                java.lang.UnsupportedOperationException: Native access is not enabled for the current module: unnamed module @1cb3ec38
                    at org.jline.terminal.impl.jni.JniTerminalProvider.checkNativeAccess(JniTerminalProvider.java:87)
                    at org.jline.terminal.impl.jni.JniTerminalProvider.<init>(JniTerminalProvider.java:67)
                    at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:62)
                    ... 4 more

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.jline/jline-terminal-jni/3.27.1/build/reports/problems/problems-report.html

5 tests completed, 5 failed

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':test'.
> There were failing tests. See the report at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.jline/jline-terminal-jni/3.27.1/build/reports/tests/test/index.html

* Try:
> Run with --scan to generate a Build Scan (Powered by Develocity).

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 1s
13 actionable tasks: 3 executed, 10 up-to-date

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':javaTest'.
> Java tests failed

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to generate a Build Scan (Powered by Develocity).
> Get more help at https://help.gradle.org.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 2s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'java run' for 3.30.7 at 2026-05-04T01:17:08Z with exit code 1 after 3s
❌ ERROR: 'java run' for 3.30.7 failed with exit code 1.
FAILED[java run][3.30.7][GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="3.30.7" ./gradlew clean javaTest -Pcoordinates="org.jline:jline-terminal-jni:3.27.1"]
```

## java run on ubuntu-latest / JDK latest-ea / mode future-defaults-all

Failure kind: java run
OS: ubuntu-latest
JDK: latest-ea
Native-image mode: future-defaults-all
Reproducer: `GVM_TCK_NATIVE_IMAGE_MODE="future-defaults-all" GVM_TCK_LV="3.30.7" ./gradlew clean javaTest -Pcoordinates="org.jline:jline-terminal-jni:3.27.1"`
Runner log: https://github.com/oracle/graalvm-reachability-metadata/actions/runs/25296069081/job/74155118887
Last 300 lines of the log:
```console
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE

> Configure project :
GraalVM Reachability Metadata TCK
---------------------------------

> Task :tckClean
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :clean

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.jline/jline-terminal-jni/3.27.1/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 748ms
9 actionable tasks: 1 executed, 8 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

> Task :clean

> Task :compileTestJava
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :compileJava NO-SOURCE
> Task :processResources NO-SOURCE
> Task :tck-test-support:compileJava UP-TO-DATE
> Task :classes UP-TO-DATE
> Task :compileTestJava

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.jline/jline-terminal-jni/3.27.1/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 955ms
10 actionable tasks: 1 executed, 9 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 2s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'javac compile' for 3.30.7 at 2026-05-04T01:20:32Z with exit code 0 after 3s
========================================================================================
 org.jline:jline-terminal-jni:3.27.1:3.30.7 stage java run
========================================================================================
Starting stage 'java run' for 3.30.7 at 2026-05-04T01:20:32Z
Command: GVM_TCK_NATIVE_IMAGE_MODE="future-defaults-all" GVM_TCK_LV="3.30.7" ./gradlew clean javaTest -Pcoordinates="org.jline:jline-terminal-jni:3.27.1"
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE

> Configure project :
GraalVM Reachability Metadata TCK
---------------------------------

> Task :tckClean
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :clean

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.jline/jline-terminal-jni/3.27.1/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 719ms
9 actionable tasks: 1 executed, 8 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

> Task :clean

> Task :javaTest
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :checkTestTimeoutAnnotations
> Task :compileJava NO-SOURCE
> Task :processResources NO-SOURCE
> Task :classes UP-TO-DATE
> Task :processTestResources NO-SOURCE
> Task :tck-test-support:compileJava UP-TO-DATE
> Task :tck-test-support:processResources NO-SOURCE
> Task :tck-test-support:classes UP-TO-DATE
> Task :tck-test-support:jar UP-TO-DATE
> Task :compileTestJava
> Task :testClasses

> Task :test FAILED

Jline_terminal_jniTest > loadsJniTerminalProviderThroughPublicServiceApi() FAILED
    java.io.IOException: Error reading provider resource file: Unable to load terminal provider jni: null
        at org.jline.terminal.spi.TerminalProvider.load(TerminalProvider.java:498)
        at org_jline.jline_terminal_jni.Jline_terminal_jniTest.loadsJniTerminalProviderThroughPublicServiceApi(Jline_terminal_jniTest.java:42)

        Caused by:
        java.io.IOException: Unable to load terminal provider jni: null
            at org.jline.terminal.spi.TerminalProvider.load(TerminalProvider.java:493)
            ... 1 more

            Caused by:
            java.lang.reflect.InvocationTargetException
                at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:74)
                at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:502)
                at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:486)
                at org.jline.terminal.spi.TerminalProvider.load(TerminalProvider.java:491)
                ... 1 more

                Caused by:
                java.lang.UnsupportedOperationException: Native access is not enabled for the current module: unnamed module @7f69d591
                    at org.jline.terminal.impl.jni.JniTerminalProvider.checkNativeAccess(JniTerminalProvider.java:87)
                    at org.jline.terminal.impl.jni.JniTerminalProvider.<init>(JniTerminalProvider.java:67)
                    at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:62)
                    ... 4 more

Jline_terminal_jniTest > terminalBuilderCreatesJniBackedTerminalForExplicitStreams() FAILED
    java.lang.IllegalStateException: Unable to create a terminal
        at org.jline.terminal.TerminalBuilder.doBuild(TerminalBuilder.java:748)
        at org.jline.terminal.TerminalBuilder.build(TerminalBuilder.java:717)
        at org_jline.jline_terminal_jni.Jline_terminal_jniTest.buildTerminal(Jline_terminal_jniTest.java:221)
        at org_jline.jline_terminal_jni.Jline_terminal_jniTest.terminalBuilderCreatesJniBackedTerminalForExplicitStreams(Jline_terminal_jniTest.java:158)

Jline_terminal_jniTest > nativePtyAppliesAttributeChangesAfterOpen() FAILED
    java.io.IOException: Error reading provider resource file: Unable to load terminal provider jni: null
        at org.jline.terminal.spi.TerminalProvider.load(TerminalProvider.java:498)
        at org_jline.jline_terminal_jni.Jline_terminal_jniTest.nativePtyAppliesAttributeChangesAfterOpen(Jline_terminal_jniTest.java:126)

        Caused by:
        java.io.IOException: Unable to load terminal provider jni: null
            at org.jline.terminal.spi.TerminalProvider.load(TerminalProvider.java:493)
            ... 1 more

            Caused by:
            java.lang.reflect.InvocationTargetException
                at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:74)
                at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:502)
                at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:486)
                at org.jline.terminal.spi.TerminalProvider.load(TerminalProvider.java:491)
                ... 1 more

                Caused by:
                java.lang.UnsupportedOperationException: Native access is not enabled for the current module: unnamed module @7f69d591
                    at org.jline.terminal.impl.jni.JniTerminalProvider.checkNativeAccess(JniTerminalProvider.java:87)
                    at org.jline.terminal.impl.jni.JniTerminalProvider.<init>(JniTerminalProvider.java:67)
                    at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:62)
                    ... 4 more

Jline_terminal_jniTest > openCreatesConfiguredNativePtyOnSupportedPosixPlatforms() FAILED
    java.io.IOException: Error reading provider resource file: Unable to load terminal provider jni: null
        at org.jline.terminal.spi.TerminalProvider.load(TerminalProvider.java:498)
        at org_jline.jline_terminal_jni.Jline_terminal_jniTest.openCreatesConfiguredNativePtyOnSupportedPosixPlatforms(Jline_terminal_jniTest.java:56)

        Caused by:
        java.io.IOException: Unable to load terminal provider jni: null
            at org.jline.terminal.spi.TerminalProvider.load(TerminalProvider.java:493)
            ... 1 more

            Caused by:
            java.lang.reflect.InvocationTargetException
                at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:74)
                at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:502)
                at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:486)
                at org.jline.terminal.spi.TerminalProvider.load(TerminalProvider.java:491)
                ... 1 more

                Caused by:
                java.lang.UnsupportedOperationException: Native access is not enabled for the current module: unnamed module @7f69d591
                    at org.jline.terminal.impl.jni.JniTerminalProvider.checkNativeAccess(JniTerminalProvider.java:87)
                    at org.jline.terminal.impl.jni.JniTerminalProvider.<init>(JniTerminalProvider.java:67)
                    at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:62)
                    ... 4 more

Jline_terminal_jniTest > nativePtyExposesConnectedMasterAndSlaveStreams() FAILED
    java.io.IOException: Error reading provider resource file: Unable to load terminal provider jni: null
        at org.jline.terminal.spi.TerminalProvider.load(TerminalProvider.java:498)
        at org_jline.jline_terminal_jni.Jline_terminal_jniTest.nativePtyExposesConnectedMasterAndSlaveStreams(Jline_terminal_jniTest.java:95)

        Caused by:
        java.io.IOException: Unable to load terminal provider jni: null
            at org.jline.terminal.spi.TerminalProvider.load(TerminalProvider.java:493)
            ... 1 more

            Caused by:
            java.lang.reflect.InvocationTargetException
                at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:74)
                at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:502)
                at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:486)
                at org.jline.terminal.spi.TerminalProvider.load(TerminalProvider.java:491)
                ... 1 more

                Caused by:
                java.lang.UnsupportedOperationException: Native access is not enabled for the current module: unnamed module @7f69d591
                    at org.jline.terminal.impl.jni.JniTerminalProvider.checkNativeAccess(JniTerminalProvider.java:87)
                    at org.jline.terminal.impl.jni.JniTerminalProvider.<init>(JniTerminalProvider.java:67)
                    at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:62)
                    ... 4 more

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.jline/jline-terminal-jni/3.27.1/build/reports/problems/problems-report.html

5 tests completed, 5 failed

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':test'.
> There were failing tests. See the report at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/org.jline/jline-terminal-jni/3.27.1/build/reports/tests/test/index.html

* Try:
> Run with --scan to generate a Build Scan (Powered by Develocity).

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 2s
13 actionable tasks: 3 executed, 10 up-to-date

> Task :javaTest FAILED

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':javaTest'.
> Java tests failed

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to generate a Build Scan (Powered by Develocity).
> Get more help at https://help.gradle.org.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 3s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'java run' for 3.30.7 at 2026-05-04T01:20:36Z with exit code 1 after 4s
❌ ERROR: 'java run' for 3.30.7 failed with exit code 1.
FAILED[java run][3.30.7][GVM_TCK_NATIVE_IMAGE_MODE="future-defaults-all" GVM_TCK_LV="3.30.7" ./gradlew clean javaTest -Pcoordinates="org.jline:jline-terminal-jni:3.27.1"]
```

--- a/metadata/org.jline/jline-terminal-jni/3.30.7/reachability-metadata.json
+++ b/metadata/org.jline/jline-terminal-jni/3.30.7/reachability-metadata.json
@@ -1,0 +1,138 @@
+{
+  "reflection": [
+    {
+      "condition": {
+        "typeReached": "org.jline.terminal.impl.jni.linux.LinuxNativePty"
+      },
+      "type": "java.io.FileDescriptor",
+      "jniAccessible": true,
+      "fields": [
+        {
+          "name": "fd"
+        }
+      ],
+      "methods": [
+        {
+          "name": "<init>",
+          "parameterTypes": []
+        }
+      ]
+    },
+    {
+      "condition": {
+        "typeReached": "org.jline.terminal.impl.jni.JniNativePty"
+      },
+      "type": "org.jline.nativ.CLibrary",
+      "jniAccessible": true,
+      "fields": [
+        {
+          "name": "TCSADRAIN"
+        },
+        {
+          "name": "TCSAFLUSH"
+        },
+        {
+          "name": "TCSANOW"
+        },
+        {
+          "name": "TIOCGWINSZ"
+        },
+        {
+          "name": "TIOCSWINSZ"
+        }
+      ]
+    },
+    {
+      "condition": {
+        "typeReached": "org.jline.terminal.impl.jni.JniTerminalProvider"
+      },
+      "type": "org.jline.nativ.CLibrary$Termios",
+      "jniAccessible": true,
+      "fields": [
+        {
+          "name": "c_cc"
+        },
+        {
+          "name": "c_cflag"
+        },
+        {
+          "name": "c_iflag"
+        },
+        {
+          "name": "c_ispeed"
+        },
+        {
+          "name": "c_lflag"
+        },
+        {
+          "name": "c_oflag"
+        },
+        {
+          "name": "c_ospeed"
+        }
+      ]
+    },
+    {
+      "condition": {
+        "typeReached": "org.jline.terminal.impl.jni.linux.LinuxNativePty"
+      },
+      "type": "org.jline.nativ.CLibrary$Termios",
+      "jniAccessible": true,
+      "fields": [
+        {
+          "name": "SIZEOF"
+        }
+      ]
+    },
+    {
+      "condition": {
+        "typeReached": "org.jline.terminal.impl.jni.JniNativePty"
+      },
+      "type": "org.jline.nativ.CLibrary$WinSize",
+      "jniAccessible": true,
+      "fields": [
+        {
+          "name": "SIZEOF"
+        },
+        {
+          "name": "ws_col"
+        },
+        {
+          "name": "ws_row"
+        },
+        {
+          "name": "ws_xpixel"
+        },
+        {
+          "name": "ws_ypixel"
+        }
+      ]
+    }
+  ],
+  "resources": [
+    {
+      "condition": {
+        "typeReached": "org.jline.terminal.impl.jni.JniNativePty"
+      },
+      "glob": "META-INF/maven/org.jline/jline-native/pom.properties"
+    },
+    {
+      "condition": {
+        "typeReached": "org.jline.terminal.impl.jni.JniNativePty"
+      },
+      "glob": "org/jline/nativ/Linux/x86_64/libjlinenative.so"
+    },
+    {
+      "condition": {
+        "typeReached": "org.jline.terminal.impl.jni.JniTerminalProvider"
+      },
+      "glob": "org/jline/utils/capabilities.txt"
+    },
+    {
+      "condition": {
+        "typeReached": "org.jline.terminal.impl.jni.JniTerminalProvider"
+      },
+      "glob": "org/jline/utils/xterm-256color.caps"
+    }
+  ]
+}

--- a/metadata/org.jline/jline-terminal-jni/index.json
+++ b/metadata/org.jline/jline-terminal-jni/index.json
@@ -1,6 +1,15 @@
 [
   {
     "latest" : true,
+    "metadata-version" : "3.30.7",
+    "tested-versions" : [
+      "3.30.7"
+    ],
+    "allowed-packages" : [
+      "org.jline.terminal.impl.jni"
+    ]
+  },
+  {
     "metadata-version" : "3.27.1",
     "source-code-url" : "https://repo.maven.apache.org/maven2/org/jline/jline-terminal-jni/$version$/jline-terminal-jni-$version$-sources.jar",
     "repository-url" : "https://github.com/jline/jline3",

--- a/metadata/org.jline/jline-terminal-jni/index.json
+++ b/metadata/org.jline/jline-terminal-jni/index.json
@@ -2,6 +2,11 @@
   {
     "latest" : true,
     "metadata-version" : "3.30.7",
+    "source-code-url" : "https://repo.maven.apache.org/maven2/org/jline/jline-terminal-jni/$version$/jline-terminal-jni-$version$-sources.jar",
+    "repository-url" : "https://github.com/jline/jline3",
+    "test-code-url" : "https://github.com/jline/jline3/tree/jline-$version$/terminal-jni/src/test/java",
+    "documentation-url" : "https://repo.maven.apache.org/maven2/org/jline/jline-terminal-jni/$version$/jline-terminal-jni-$version$-javadoc.jar",
+    "description" : "JLine JNI Terminal provides the JLine terminal provider implementation that uses JLine's JNI native library for low-level terminal and pseudoterminal access. It plugs into the JLine terminal API to support native console operations across supported platforms without relying on JNA or external process execution.",
     "tested-versions" : [
       "3.30.7"
     ],

--- a/stats/org.jline/jline-terminal-jni/3.30.7/execution-metrics.json
+++ b/stats/org.jline/jline-terminal-jni/3.30.7/execution-metrics.json
@@ -1,0 +1,93 @@
+{
+  "fix_java_run_fail:2026-05-04": {
+    "timestamp": "2026-05-04T03:36:52.326718Z",
+    "library": "org.jline:jline-terminal-jni:3.30.7",
+    "starting_commit": "de4c066cc2fa8b508967ce7c0db9145bbfeb46a1",
+    "ending_commit": "adc9a7da387af59a492885488a7e569d013ca354",
+    "previous_library": "org.jline:jline-terminal-jni:3.27.1",
+    "strategy_name": "java_run_iterative_with_coverage_sources_pi_gpt-5.5",
+    "agent": "pi",
+    "model": "oca/gpt-5.5",
+    "status": "success",
+    "stats": {
+      "version": "3.30.7",
+      "dynamicAccess": {
+        "breakdown": {
+          "reflection": {
+            "coverageRatio": 1.0,
+            "coveredCalls": 3,
+            "totalCalls": 3
+          }
+        },
+        "coverageRatio": 1.0,
+        "coveredCalls": 3,
+        "totalCalls": 3
+      },
+      "libraryCoverage": {
+        "instruction": {
+          "covered": 1501,
+          "missed": 6093,
+          "ratio": 0.197656,
+          "total": 7594
+        },
+        "line": {
+          "covered": 243,
+          "missed": 961,
+          "ratio": 0.201827,
+          "total": 1204
+        },
+        "method": {
+          "covered": 39,
+          "missed": 92,
+          "ratio": 0.29771,
+          "total": 131
+        }
+      }
+    },
+    "previous_library_stats": {
+      "version": "3.27.1",
+      "dynamicAccess": {
+        "breakdown": {},
+        "coverageRatio": 1.0,
+        "coveredCalls": 0,
+        "totalCalls": 0
+      },
+      "libraryCoverage": {
+        "instruction": {
+          "covered": 1461,
+          "missed": 5892,
+          "ratio": 0.198694,
+          "total": 7353
+        },
+        "line": {
+          "covered": 231,
+          "missed": 935,
+          "ratio": 0.198113,
+          "total": 1166
+        },
+        "method": {
+          "covered": 37,
+          "missed": 84,
+          "ratio": 0.305785,
+          "total": 121
+        }
+      }
+    },
+    "metrics": {
+      "input_tokens_used": 30354,
+      "cached_input_tokens_used": 291328,
+      "output_tokens_used": 2815,
+      "iterations": 1,
+      "cost_usd": 0.2301,
+      "tested_library_loc": 243,
+      "metadata_entries": 34,
+      "previous_library_metadata_entries": 34,
+      "code_coverage_percent": 20.18,
+      "previous_library_coverage_percent": 19.81
+    },
+    "artifacts": {
+      "test_file": "tests/src/org.jline/jline-terminal-jni/3.30.7/src/test/java/org_jline/jline_terminal_jni/Jline_terminal_jniTest.java",
+      "metadata_file": "metadata/org.jline/jline-terminal-jni/3.30.7/reachability-metadata.json"
+    }
+  }
+}

--- a/stats/org.jline/jline-terminal-jni/3.30.7/stats.json
+++ b/stats/org.jline/jline-terminal-jni/3.30.7/stats.json
@@ -1,0 +1,37 @@
+{
+  "versions" : [ {
+    "version" : "3.30.7",
+    "dynamicAccess" : {
+      "breakdown" : {
+        "reflection" : {
+          "coverageRatio" : 1.0,
+          "coveredCalls" : 3,
+          "totalCalls" : 3
+        }
+      },
+      "coverageRatio" : 1.0,
+      "coveredCalls" : 3,
+      "totalCalls" : 3
+    },
+    "libraryCoverage" : {
+      "instruction" : {
+        "covered" : 1501,
+        "missed" : 6093,
+        "ratio" : 0.197656,
+        "total" : 7594
+      },
+      "line" : {
+        "covered" : 243,
+        "missed" : 961,
+        "ratio" : 0.201827,
+        "total" : 1204
+      },
+      "method" : {
+        "covered" : 39,
+        "missed" : 92,
+        "ratio" : 0.29771,
+        "total" : 131
+      }
+    }
+  } ]
+}

--- a/tests/src/org.jline/jline-terminal-jni/3.30.7/.gitignore
+++ b/tests/src/org.jline/jline-terminal-jni/3.30.7/.gitignore
@@ -1,0 +1,4 @@
+gradlew.bat
+gradlew
+gradle/
+build/

--- a/tests/src/org.jline/jline-terminal-jni/3.30.7/build.gradle
+++ b/tests/src/org.jline/jline-terminal-jni/3.30.7/build.gradle
@@ -1,0 +1,27 @@
+/*
+ * Copyright and related rights waived via CC0
+ *
+ * You should have received a copy of the CC0 legalcode along with this
+ * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+ */
+plugins {
+    id "org.graalvm.internal.tck"
+}
+
+String libraryVersion = tck.testedLibraryVersion.get()
+
+dependencies {
+    testImplementation "org.jline:jline-terminal-jni:$libraryVersion"
+    testImplementation 'org.assertj:assertj-core:3.22.0'
+}
+
+graalvmNative {
+    agent {
+        defaultMode = "conditional"
+        modes {
+            conditional {
+                userCodeFilterPath = "user-code-filter.json"
+            }
+        }
+    }
+}

--- a/tests/src/org.jline/jline-terminal-jni/3.30.7/build.gradle
+++ b/tests/src/org.jline/jline-terminal-jni/3.30.7/build.gradle
@@ -15,6 +15,12 @@ dependencies {
     testImplementation 'org.assertj:assertj-core:3.22.0'
 }
 
+tasks.withType(Test).configureEach {
+    jvmArgs += [
+        "--enable-native-access=ALL-UNNAMED"
+    ]
+}
+
 graalvmNative {
     agent {
         defaultMode = "conditional"
@@ -22,6 +28,13 @@ graalvmNative {
             conditional {
                 userCodeFilterPath = "user-code-filter.json"
             }
+        }
+    }
+    binaries {
+        all {
+            buildArgs.addAll([
+                '--enable-native-access=ALL-UNNAMED'
+            ])
         }
     }
 }

--- a/tests/src/org.jline/jline-terminal-jni/3.30.7/gradle.properties
+++ b/tests/src/org.jline/jline-terminal-jni/3.30.7/gradle.properties
@@ -1,0 +1,6 @@
+# Optional explicit tested library coordinates (format: group:artifact:version).
+# If omitted, fallback resolution in org.graalvm.internal.tck.gradle still uses
+# environment variable GVM_TCK_LC and then this property.
+library.coordinates = org.jline:jline-terminal-jni:3.30.7
+library.version = 3.30.7
+metadata.dir = org.jline/jline-terminal-jni/3.30.7/

--- a/tests/src/org.jline/jline-terminal-jni/3.30.7/settings.gradle
+++ b/tests/src/org.jline/jline-terminal-jni/3.30.7/settings.gradle
@@ -1,0 +1,20 @@
+/*
+ * Copyright and related rights waived via CC0
+ *
+ * You should have received a copy of the CC0 legalcode along with this
+ * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+ */
+pluginManagement {
+    def tckPath = Objects.requireNonNullElse(
+            System.getenv("GVM_TCK_TCKDIR"),
+            "../../../../tck-build-logic"
+    )
+    includeBuild(tckPath)
+    repositories { mavenLocal(); gradlePluginPortal() }
+}
+
+plugins {
+    id "org.graalvm.internal.tck-settings" version "1.0.0-SNAPSHOT"
+}
+
+rootProject.name = 'org.jline.jline-terminal-jni_tests'

--- a/tests/src/org.jline/jline-terminal-jni/3.30.7/src/test/java/org_jline/jline_terminal_jni/Jline_terminal_jniTest.java
+++ b/tests/src/org.jline/jline-terminal-jni/3.30.7/src/test/java/org_jline/jline_terminal_jni/Jline_terminal_jniTest.java
@@ -1,0 +1,240 @@
+/*
+ * Copyright and related rights waived via CC0
+ *
+ * You should have received a copy of the CC0 legalcode along with this
+ * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+ */
+package org_jline.jline_terminal_jni;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.EOFException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.Set;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+
+import org.jline.terminal.Attributes;
+import org.jline.terminal.Size;
+import org.jline.terminal.Terminal;
+import org.jline.terminal.TerminalBuilder;
+import org.jline.terminal.impl.jni.JniNativePty;
+import org.jline.terminal.impl.jni.JniTerminalProvider;
+import org.jline.terminal.spi.Pty;
+import org.jline.terminal.spi.SystemStream;
+import org.jline.terminal.spi.TerminalExt;
+import org.jline.terminal.spi.TerminalProvider;
+import org.junit.jupiter.api.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+public class Jline_terminal_jniTest {
+    private static final Set<String> SUPPORTED_POSIX_OS_PREFIXES = Set.of(
+            "Linux", "Mac", "Darwin", "Solaris", "SunOS", "FreeBSD");
+
+    @Test
+    void loadsJniTerminalProviderThroughPublicServiceApi() throws Exception {
+        TerminalProvider provider = TerminalProvider.load("jni");
+
+        assertThat(provider).isInstanceOf(JniTerminalProvider.class);
+        assertThat(provider.name()).isEqualTo("jni");
+        assertThat(provider).hasToString("TerminalProvider[jni]");
+
+        for (SystemStream stream : SystemStream.values()) {
+            assertThat(provider.systemStreamWidth(stream)).isGreaterThanOrEqualTo(-1);
+            assertThat(provider.isSystemStream(stream)).isIn(true, false);
+        }
+    }
+
+    @Test
+    void openCreatesConfiguredNativePtyOnSupportedPosixPlatforms() throws Exception {
+        TerminalProvider provider = TerminalProvider.load("jni");
+        Attributes attributes = attributesForPty();
+        Size requestedSize = new Size(81, 27);
+
+        if (isSupportedPosixOperatingSystem()) {
+            try (Pty pty = ((JniTerminalProvider) provider).open(attributes, requestedSize)) {
+                assertThat(pty).isInstanceOf(JniNativePty.class);
+                assertThat(pty.getProvider()).isSameAs(provider);
+                assertThat(pty.getSystemStream()).isNull();
+
+                JniNativePty nativePty = (JniNativePty) pty;
+                assertThat(nativePty.getMaster()).isGreaterThan(0);
+                assertThat(nativePty.getSlave()).isGreaterThan(0);
+                assertThat(nativePty.getSlaveOut()).isEqualTo(nativePty.getSlave());
+                assertThat(nativePty.getMasterFD()).isNotNull();
+                assertThat(nativePty.getSlaveFD()).isNotNull();
+                assertThat(nativePty.getSlaveOutFD()).isSameAs(nativePty.getSlaveFD());
+                assertThat(nativePty.getName()).isNotBlank();
+                assertThat(nativePty).hasToString("NativePty[" + nativePty.getName() + "]");
+
+                assertThat(nativePty.getSize()).isEqualTo(requestedSize);
+                Size resized = new Size(100, 31);
+                nativePty.setSize(resized);
+                assertThat(nativePty.getSize()).isEqualTo(resized);
+
+                Attributes actualAttributes = nativePty.getAttr();
+                assertThat(actualAttributes.getLocalFlag(Attributes.LocalFlag.ECHO)).isFalse();
+                assertThat(actualAttributes.getLocalFlag(Attributes.LocalFlag.ICANON)).isTrue();
+                assertThat(actualAttributes.getInputFlag(Attributes.InputFlag.ICRNL)).isTrue();
+                assertThat(actualAttributes.getControlChar(Attributes.ControlChar.VERASE)).isEqualTo(127);
+            }
+        } else {
+            assertThatThrownBy(() -> ((JniTerminalProvider) provider).open(attributes, requestedSize))
+                    .isInstanceOf(UnsupportedOperationException.class);
+        }
+    }
+
+    @Test
+    void nativePtyExposesConnectedMasterAndSlaveStreams() throws Exception {
+        TerminalProvider provider = TerminalProvider.load("jni");
+        Attributes attributes = attributesForPty();
+        Size requestedSize = new Size(80, 24);
+
+        if (isSupportedPosixOperatingSystem()) {
+            try (Pty pty = ((JniTerminalProvider) provider).open(attributes, requestedSize)) {
+                byte[] payload = "jni-pty-payload".getBytes(StandardCharsets.UTF_8);
+                InputStream masterInput = pty.getMasterInput();
+                OutputStream slaveOutput = pty.getSlaveOutput();
+                ExecutorService executor = newDaemonSingleThreadExecutor("jni-pty-reader");
+                Future<byte[]> readFromMaster = executor.submit(() -> readExactly(masterInput, payload.length));
+
+                try {
+                    slaveOutput.write(payload);
+                    slaveOutput.flush();
+
+                    assertThat(readFromMaster.get(5, TimeUnit.SECONDS)).containsExactly(payload);
+                } finally {
+                    readFromMaster.cancel(true);
+                    executor.shutdownNow();
+                    assertThat(executor.awaitTermination(5, TimeUnit.SECONDS)).isTrue();
+                }
+            }
+        } else {
+            assertThatThrownBy(() -> ((JniTerminalProvider) provider).open(attributes, requestedSize))
+                    .isInstanceOf(UnsupportedOperationException.class);
+        }
+    }
+
+    @Test
+    void nativePtyAppliesAttributeChangesAfterOpen() throws Exception {
+        TerminalProvider provider = TerminalProvider.load("jni");
+        Attributes attributes = attributesForPty();
+        Size requestedSize = new Size(80, 24);
+
+        if (isSupportedPosixOperatingSystem()) {
+            try (Pty pty = ((JniTerminalProvider) provider).open(attributes, requestedSize)) {
+                Attributes updatedAttributes = new Attributes(pty.getAttr());
+                updatedAttributes.setLocalFlag(Attributes.LocalFlag.ECHO, true);
+                updatedAttributes.setLocalFlag(Attributes.LocalFlag.ICANON, false);
+                updatedAttributes.setInputFlag(Attributes.InputFlag.ICRNL, false);
+                updatedAttributes.setControlChar(Attributes.ControlChar.VMIN, 0);
+                updatedAttributes.setControlChar(Attributes.ControlChar.VTIME, 1);
+
+                pty.setAttr(updatedAttributes);
+
+                Attributes actualAttributes = pty.getAttr();
+                assertThat(actualAttributes.getLocalFlag(Attributes.LocalFlag.ECHO)).isTrue();
+                assertThat(actualAttributes.getLocalFlag(Attributes.LocalFlag.ICANON)).isFalse();
+                assertThat(actualAttributes.getInputFlag(Attributes.InputFlag.ICRNL)).isFalse();
+                assertThat(actualAttributes.getControlChar(Attributes.ControlChar.VMIN)).isZero();
+                assertThat(actualAttributes.getControlChar(Attributes.ControlChar.VTIME)).isEqualTo(1);
+            }
+        } else {
+            assertThatThrownBy(() -> ((JniTerminalProvider) provider).open(attributes, requestedSize))
+                    .isInstanceOf(UnsupportedOperationException.class);
+        }
+    }
+
+    @Test
+    void terminalBuilderCreatesJniBackedTerminalForExplicitStreams() throws Exception {
+        if (isSupportedPosixOperatingSystem()) {
+            ByteArrayOutputStream output = new ByteArrayOutputStream();
+            try (Terminal terminal = buildTerminal(output)) {
+                assertThat(terminal.getName()).isEqualTo("jni-test-terminal");
+                assertThat(terminal.getType()).isEqualTo("xterm-256color");
+                assertThat(terminal.encoding()).isEqualTo(StandardCharsets.UTF_8);
+                assertThat(terminal).isInstanceOf(TerminalExt.class);
+
+                TerminalExt terminalExt = (TerminalExt) terminal;
+                assertThat(terminalExt.getProvider().name()).isEqualTo("jni");
+                assertThat(terminalExt.getSystemStream()).isNull();
+
+                assertThat(terminal.getSize()).isEqualTo(new Size(90, 33));
+                terminal.setSize(new Size(91, 34));
+                assertThat(terminal.getWidth()).isEqualTo(91);
+                assertThat(terminal.getHeight()).isEqualTo(34);
+
+                assertThat(terminal.echo()).isFalse();
+                assertThat(terminal.canPauseResume()).isTrue();
+                assertThat(terminal.paused()).isTrue();
+                terminal.pause();
+                assertThat(terminal.paused()).isTrue();
+                assertThat(terminal.reader()).isNotNull();
+                assertThat(terminal.writer()).isNotNull();
+                assertThat(output.size()).isZero();
+            }
+        } else {
+            assertThatThrownBy(() -> buildTerminal(new ByteArrayOutputStream()))
+                    .isInstanceOf(UnsupportedOperationException.class);
+        }
+    }
+
+    private static ExecutorService newDaemonSingleThreadExecutor(String threadName) {
+        return Executors.newSingleThreadExecutor(runnable -> {
+            Thread thread = new Thread(runnable, threadName);
+            thread.setDaemon(true);
+            return thread;
+        });
+    }
+
+    private static byte[] readExactly(InputStream input, int length) throws Exception {
+        byte[] buffer = new byte[length];
+        int offset = 0;
+        while (offset < length) {
+            int read = input.read(buffer, offset, length - offset);
+            if (read < 0) {
+                throw new EOFException("Expected " + length + " bytes, read " + offset);
+            }
+            offset += read;
+        }
+        return buffer;
+    }
+
+    private static Terminal buildTerminal(ByteArrayOutputStream output) throws Exception {
+        return TerminalBuilder.builder()
+                .provider("jni")
+                .system(false)
+                .name("jni-test-terminal")
+                .type("xterm-256color")
+                .encoding(StandardCharsets.UTF_8)
+                .streams(new ByteArrayInputStream(new byte[0]), output)
+                .attributes(attributesForPty())
+                .size(new Size(90, 33))
+                .nativeSignals(false)
+                .paused(true)
+                .build();
+    }
+
+    private static Attributes attributesForPty() {
+        Attributes attributes = new Attributes();
+        attributes.setLocalFlag(Attributes.LocalFlag.ECHO, false);
+        attributes.setLocalFlag(Attributes.LocalFlag.ICANON, true);
+        attributes.setInputFlag(Attributes.InputFlag.ICRNL, true);
+        attributes.setOutputFlag(Attributes.OutputFlag.OPOST, true);
+        attributes.setControlFlag(Attributes.ControlFlag.CREAD, true);
+        attributes.setControlChar(Attributes.ControlChar.VERASE, 127);
+        attributes.setControlChar(Attributes.ControlChar.VMIN, 1);
+        return attributes;
+    }
+
+    private static boolean isSupportedPosixOperatingSystem() {
+        String osName = System.getProperty("os.name", "");
+        return SUPPORTED_POSIX_OS_PREFIXES.stream().anyMatch(osName::startsWith);
+    }
+}

--- a/tests/src/org.jline/jline-terminal-jni/3.30.7/user-code-filter.json
+++ b/tests/src/org.jline/jline-terminal-jni/3.30.7/user-code-filter.json
@@ -1,0 +1,10 @@
+{
+  "rules" : [
+    {
+      "excludeClasses" : "**"
+    },
+    {
+      "includeClasses" : "org.jline.terminal.impl.jni.**"
+    }
+  ]
+}

--- a/tests/src/org.jline/jline-terminal-jni/3.30.7/user-code-filter.json
+++ b/tests/src/org.jline/jline-terminal-jni/3.30.7/user-code-filter.json
@@ -5,6 +5,9 @@
     },
     {
       "includeClasses" : "org.jline.terminal.impl.jni.**"
+    },
+    {
+      "includeClasses" : "org_jline.jline_terminal_jni.**"
     }
   ]
 }

